### PR TITLE
platform/connectors

### PR DESCRIPTION
**Pull request by sahils-tenable**
- Adds support for version 3 connectors endpoints.

**Todo:**
- Need to be moved under platform as `tio.v3.platform.connectors`
- property for `platforms` is defined in #81 
- Once #81 is merged we can pull from develop and update code structure to follow `tio.v3.platform.connectors`

### CODE REVIEW

File: tenable/io/v3/connectors/api.py [hunk 1]
```diff
@@ -0,0 +1,389 @@
...
+                The name of the connector to create.
+            con_type:
+                The type of connector to create.
+                Example:
+                    'aws_keyless'
+            schedule:
+                The data import schedule
+                Example:
+                    ('days', 2)
+            network_id:
```
Comment (harshshah-tenable) [inline tenable/io/v3/connectors/api.py]: Please add type in params in docstrings?
i.e name (str)
con_type (str)
schedule (dict, optional)
network_id (str, optional)

File: tenable/io/v3/connectors/api.py [hunk 2]
```diff
@@ -0,0 +1,389 @@
...
+                communicate with the AWS API, including access_key
+                and secret_key.
+
+                Examples:
+                    {
+                    'access_key':
+                    'd08c2a75517ad18da1d1d2e8f9ea70792f658674b352a9a00e7fcc4a74dcff6a',  # noqa E501
+                    'secret_key':
+                    'be13082452f57880261e8170290490026417745b8cde8a7d1e554d16eff690bf'  # noqa E501
+                    }
```
Comment (harshshah-tenable) [inline tenable/io/v3/connectors/api.py]:
```suggestion
               >>>     {
               ...     'access_key':
               ...     'd08c2a75517ad18da1d1d2e8f9ea70792f658674b352a9a00e7fcc4a74dcff6a',  # noqa E501
               ...     'secret_key':
               ...     'be13082452f57880261e8170290490026417745b8cde8a7d1e554d16eff690bf'  # noqa E501
               ...     }
```

File: tenable/io/v3/connectors/api.py [hunk 3]
```diff
@@ -0,0 +1,389 @@
...
+        return self._post('aws/cloudtrails', json=payload)['trails']
+
+    def delete(self, connector_id: UUID) -> None:
+        '''
+        Deletes a connectors.
+
+        :devportal:`connectors: delete <connectors-delete>`
+
+        Args:
+            connector_id :
```
Comment (harshshah-tenable) [inline tenable/io/v3/connectors/api.py]:
```suggestion
            connector_id (uuid):
```

File: tenable/io/v3/connectors/api.py (hunk 1 above)
Comment (sahils-tenable) [reply tenable/io/v3/connectors/api.py]: are we adding type params at both places? In function constructor and in docstring?

Comment (harshshah-tenable) [reply tenable/io/v3/connectors/api.py]: Yes

File: tenable/io/v3/connectors/api.py [hunk 4]
```diff
@@ -0,0 +1,391 @@
...
+        Returns:
+            :obj:`Dict`:
+                The connector resource record.
+
+        Example:
+            >>> resp = tio.v3.connectors.create(
+        ... name='test', con_type='aws_keyless',
+        ... network_id='00000000-0000-0000-0000-000000000000',
+        ... schedule=('days', 2), sub_accounts=[{}]
+        ... )
```
Comment (harshshah-tenable) [inline tenable/io/v3/connectors/api.py]: Please add an indentation here

File: tests/io/v3/connectors/test_connectors_api.py [hunk 5]
```diff
@@ -0,0 +1,584 @@
...
+                ')]',
+            'logicAppName': 'TenableFAEventBasedCollection',
+            'resourceTags': {
+                'Tenable': 'AzureFa'
+            },
+            'roleDefinition': {
+                'name':
+                    '[format("Tenable FA Role (Subscription: {0} |'
+                    ' Connector: {1})", subscription().subscriptionId, '
+                    'variables("connectorId"))]',
```
Comment (harshshah-tenable) [inline tests/io/v3/connectors/test_connectors_api.py:239]: Please replace " -> ' at all instances

**Review by harshshah-tenable — APPROVED** [review]

**Comment by harshshah-tenable** [comment]
please move connectors code inside platforms module, `tio.v3.platforms.connectors`

File: tests/io/v3/connectors/test_connectors_api.py (hunk 5 above)
Comment (sahils-tenable) [reply tests/io/v3/connectors/test_connectors_api.py:239]: I guess we are using single quotes at all place.

File: tenable/io/v3/connectors/api.py [hunk 6]
```diff
@@ -0,0 +1,393 @@
+'''
+Connectors
+==========
+
+The following methods allow for interaction into the Tenable.io
```
Comment (Mangusingh-te) [inline tenable/io/v3/connectors/api.py:5]:
```suggestion
This module contains ConnectorsAPI class, that allows interaction with Tenable.io
```

File: tenable/io/v3/connectors/api.py [hunk 7]
```diff
@@ -0,0 +1,393 @@
...
+            'account_id': account_id
+        }
+        payload = dict_clean(payload)
+        schema = ConnectorListTrails()
+        payload = schema.dump(schema.load(payload))
+        return self._post('aws/cloudtrails', json=payload)['trails']
+
+    def delete(self, connector_id: UUID) -> None:
+        '''
+        Deletes a connectors.
```
Comment (Mangusingh-te) [inline tenable/io/v3/connectors/api.py]:
```suggestion
        Deletes a connector.
```

File: tenable/io/v3/connectors/api.py [hunk 8]
```diff
@@ -0,0 +1,393 @@
...
+
+        :devportal:`connectors: delete <connectors-delete>`
+
+        Args:
+            connector_id (UUID):
+                The unique identifier for the connectors to delete.
```
Comment (Mangusingh-te) [inline tenable/io/v3/connectors/api.py]:
```suggestion
                The unique identifier for the connector to delete.
```

File: tenable/io/v3/connectors/api.py [hunk 9]
```diff
@@ -0,0 +1,393 @@
...
+            'network_id': network_id,
+            'params': params,
+            'schedule': schedule
+
+        }
+        payload = dict_clean(payload)
+        schema = ConnectorCreate()
+        payload = schema.dump(schema.load(payload))
+
+        connector_info = self.details(connector_id)
```
Comment (Mangusingh-te) [inline tenable/io/v3/connectors/api.py:292]: Why are we querying connector details in **edit** operation ? Does update API endpoint supports partial updates ?

File: tenable/io/v3/connectors/api.py [hunk 10]
```diff
@@ -0,0 +1,393 @@
...
+                Examples:
+                    >>> [('field_name_1', 'asc'),
+                    ...      ('field_name_2', 'desc')]
+            limit (int, optional):
+                Number of objects to be returned in each request.
+                Default limit is 200.
+            next (str, optional):
+                The pagination token to use when requesting the next page of
+                results. This token is presented in the previous response.
+            return_resp (bool, optional):
```
Comment (Mangusingh-te) [inline tenable/io/v3/connectors/api.py:364]: Update docstring.

File: tenable/io/v3/connectors/api.py [hunk 11]
```diff
@@ -0,0 +1,393 @@
...
+                specific page of data.
+            return_csv (bool, optional):
+                If set to true, it will return the CSV response or
+                iterable (based on return_resp flag). Iterator returns all
+                rows in text/csv format for each call with row headers.
+        Returns:
+            Iterable:
+                The iterable that handles the pagination for the job.
+            requests.Response:
+                If ``return_json`` was set to ``True``, then a response
```
Comment (Mangusingh-te) [inline tenable/io/v3/connectors/api.py]: Update docstring.

File: tenable/io/v3/connectors/schema.py [hunk 12]
```diff
@@ -0,0 +1,48 @@
+from marshmallow import Schema, ValidationError, fields, pre_load
+from marshmallow import validate as v
+
+
+class ConnectorSchedule(Schema):
+    units = fields.Str()
```
Comment (Mangusingh-te) [inline tenable/io/v3/connectors/schema.py:6]: Shouldn't it be **unit** instead of **units** ?

File: tenable/io/v3/connectors/schema.py [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+            units = data[0]
+            value = data[1]
+            return dict(units=units, value=value)
+        elif isinstance(data, dict) and len(data) == 2:
+            return data
+        else:
+            raise ValidationError('Invalid Schedule definition')
+
+
+class ConnectorCreate(Schema):
```
Comment (Mangusingh-te) [inline tenable/io/v3/connectors/schema.py]: This schema class is used during both create & edit operation. Can it be renamed to **ConnectorCreateOrEditSchema** ?

File: tests/io/v3/connectors/test_connectors_api.py [hunk 14]
```diff
@@ -0,0 +1,584 @@
...
+
+    iterator = api.v3.connectors.search(
+        fields=fields, limit=200, sort=sort, filter=filters
+    )
+    assert isinstance(iterator, SearchIterator)
+
+    event_list = []
+    for event in iterator:
+        event_list.append(event)
+    assert len(event_list) == api_response['pagination']['total']
```
Comment (Mangusingh-te) [inline tests/io/v3/connectors/test_connectors_api.py]: Refactor code to get rid of loop.

**Review by Mangusingh-te — CHANGES_REQUESTED** [review]

File: tenable/io/v3/connectors/api.py (hunk 7 above)
Comment (alpesh-te) [reply tenable/io/v3/connectors/api.py]: Addressed

File: tenable/io/v3/connectors/api.py (hunk 9 above)
Comment (alpesh-te) [reply tenable/io/v3/connectors/api.py:292]: Look like API endpoint support partial update and reset of feed by details API. we will change while writing integration test cases if API doesn't support partial parameter

File: tenable/io/v3/connectors/api.py (hunk 10 above)
Comment (alpesh-te) [reply tenable/io/v3/connectors/api.py:364]: Addressed

File: tenable/io/v3/connectors/api.py (hunk 11 above)
Comment (alpesh-te) [reply tenable/io/v3/connectors/api.py]: Addressed

File: tenable/io/v3/connectors/api.py (hunk 8 above)
Comment (alpesh-te) [reply tenable/io/v3/connectors/api.py]: Addressed

File: tenable/io/v3/connectors/schema.py (hunk 13 above)
Comment (alpesh-te) [reply tenable/io/v3/connectors/schema.py]: Addressed

File: tests/io/v3/connectors/test_connectors_api.py (hunk 14 above)
Comment (alpesh-te) [reply tests/io/v3/connectors/test_connectors_api.py]: Addressed

File: tenable/io/v3/connectors/schema.py (hunk 12 above)
Comment (alpesh-te) [reply tenable/io/v3/connectors/schema.py:6]: I just seen, it is used many places as a same key `units`, I will prefer use same everywhere

**Review by alpesh-te — APPROVED** [review]
LGTM!

File: tenable/io/v3/connectors/api.py (hunk 6 above)
Comment (alpesh-te) [reply tenable/io/v3/connectors/api.py:5]: Looks like it followed same patter on all classes. I would like to keep that same.

**Review by Mangusingh-te — APPROVED** [review]

**Pull request merged**